File: benchmarks/primes_parallel/java/Primes_parallel.java
public class Primes_parallel {
    static final int NUM_THREADS = 4;

    static boolean isPrime(long x) {
        if (x < 2) return false;
        if (x % 2 == 0) return x == 2;
        for (long d = 3; d * d <= x; d += 2) {
            if (x % d == 0) return false;
        }
        return true;
    }

    public static void main(String[] args) throws InterruptedException {
        long n = args.length > 0 ? Long.parseLong(args[0]) : 2_000_000;

        if (!isPrime(2) || !isPrime(97) || isPrime(100) || isPrime(1)) {
            System.err.println("self-check failed: is_prime disagrees with known facts");
            System.exit(1);
        }

        long chunk = n / NUM_THREADS;
        Thread[] threads = new Thread[NUM_THREADS];
        long[] counts = new long[NUM_THREADS];
        for (int i = 0; i < NUM_THREADS; i++) {
            final int idx = i;
            final long start = Math.max(idx * chunk, 2);
            final long end = (idx == NUM_THREADS - 1) ? n : (idx + 1) * chunk;
            threads[idx] = new Thread(() -> {
                long count = 0;
                for (long x = start; x < end; x++) {
                    if (isPrime(x)) count++;
                }
                counts[idx] = count;
            });
            threads[idx].start();
        }

        long total = 0;
        for (int i = 0; i < NUM_THREADS; i++) {
            threads[i].join();
            total += counts[i];
        }

        System.out.println(total);
    }
}

File: benchmarks/primes_parallel/rust/primes_parallel.rs
use std::env;
use std::thread;

const NUM_THREADS: i64 = 4;

fn is_prime(x: i64) -> bool {
    if x < 2 {
        return false;
    }
    if x % 2 == 0 {
        return x == 2;
    }
    let mut d = 3;
    while d * d <= x {
        if x % d == 0 {
            return false;
        }
        d += 2;
    }
    true
}

fn main() {
    let n: i64 = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(2_000_000);

    assert!(
        is_prime(2) && is_prime(97) && !is_prime(100) && !is_prime(1),
        "self-check failed: is_prime disagrees with known facts"
    );

    let chunk = n / NUM_THREADS;
    let mut handles = Vec::new();
    for i in 0..NUM_THREADS {
        let start = (i * chunk).max(2);
        let end = if i == NUM_THREADS - 1 { n } else { (i + 1) * chunk };
        handles.push(thread::spawn(move || {
            let mut count = 0i64;
            for x in start..end {
                if is_prime(x) {
                    count += 1;
                }
            }
            count
        }));
    }

    let total: i64 = handles.into_iter().map(|h| h.join().unwrap()).sum();
    println!("{}", total);
}
